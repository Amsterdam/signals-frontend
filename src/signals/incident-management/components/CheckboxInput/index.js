import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Label from '../Label';

const CheckboxInput = ({
  name,
  display,
  values,
  toggleChecked,
  allChecked,
  hasToggleAll,
  toggleLabel,
}) => {
  const render = ({ handler }) => {
    const inputValues = handler().value || [];
    debugger;

    return (
      <Fragment key={Math.random()}>
        {display && <Label htmlFor={`${name}-${values[0].key}`}>{display}</Label>}

        {hasToggleAll && toggleLabel && (
          <button type="button" onClick={toggleChecked}>
            {toggleLabel}
          </button>
        )}

        {values.map(({ key, value }) => (
          <div className="antwoord" key={key}>
            <input
              id={`${name}-${key}`}
              name={name}
              defaultChecked={allChecked || inputValues.includes(key)}
              type="checkbox"
              value={key}
            />
            <label htmlFor={`${name}-${key}`}>{value}</label>
          </div>
        ))}
      </Fragment>
    );
  };

  render.defaultProps = {
    touched: false,
  };

  render.propTypes = {
    handler: PropTypes.func.isRequired,
  };

  return render;
};

CheckboxInput.defaultProps = {
  hasToggleAll: true,
  toggleLabel: '',
};

CheckboxInput.propTypes = {
  /** When provided, will render a <label> element with the for attribute pointing to the first selectable element in the list */
  display: PropTypes.string,
  /** When true, will render a toggle that (de)selects all checkboxes */
  hasToggleAll: PropTypes.bool,
  /** Name of the group of elements  */
  name: PropTypes.string.isRequired,
  /** Flag indicating if all boxes should be checked or not */
  toggleChecked: PropTypes.bool,
  /** Element to show as select-all toggle label */
  toggleLabel: PropTypes.node,
  values: PropTypes.shape({
    key: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
  }).isRequired,
};

export default CheckboxInput;
