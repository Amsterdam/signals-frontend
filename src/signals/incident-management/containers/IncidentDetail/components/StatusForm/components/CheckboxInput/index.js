import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Label, themeSpacing } from '@datapunt/asc-ui';
import Checkbox from 'components/Checkbox';

const CheckboxStyle = styled.div`
  display: flex;
  margin-bottom: ${themeSpacing(4)};

  & > label > span {
    margin: 0;
  }
`;

const CheckboxInput = ({ name, label, ...rest }) => ({ handler }) => (
  <CheckboxStyle>
    <Checkbox id={name} name={name} checked={handler().value} onChange={handler().onChange} {...rest} />
    <Label htmlFor={name} label={label} />
  </CheckboxStyle>
);

CheckboxInput.propTypes = {
  name: PropTypes.string,
  label: PropTypes.string,
};

export default CheckboxInput;
