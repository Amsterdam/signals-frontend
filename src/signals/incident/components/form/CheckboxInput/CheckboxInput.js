// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2018 - 2022 Gemeente Amsterdam
import { Label } from '@amsterdam/asc-ui'
import isObject from 'lodash/isObject'
import map from 'lodash/map'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Checkbox from 'components/Checkbox'
import FormField from 'components/FormField'

function updateIncidentCheckboxMulti(
  checked,
  value,
  key,
  oldValue,
  meta,
  parent
) {
  let output = [...oldValue]
  if (checked) {
    output.push({
      id: key,
      label: value,
    })
  } else {
    output = output.filter((item) => item.id !== key)
  }

  parent.meta.updateIncident({ [meta.name]: output })
}

const CheckboxWrapper = styled.div`
  & > :first-child {
    margin-top: -6px; /* This margin enforces the 12px design distance from the title description */
  }
`

const CheckboxGroup = styled.div`
  display: flex;
  flex-direction: column;
`

const CheckboxInput = ({
  handler,
  hasError,
  meta,
  parent,
  getError,
  validatorsOrOpts,
}) =>
  meta.isVisible && (
    <FormField
      meta={meta}
      options={validatorsOrOpts}
      hasError={hasError}
      getError={getError}
    >
      <CheckboxWrapper>
        {isObject(meta.values) ? (
          <CheckboxGroup role="group" id={meta.name}>
            <input type="hidden" {...handler()} />

            {map(meta.values, (value, key) => (
              <Label
                htmlFor={`${meta.name}-${key + 1}`}
                label={value}
                key={`${meta.name}-${key + 1}`}
                noActiveState
              >
                <Checkbox
                  id={`${meta.name}-${key + 1}`}
                  name={`${meta.name}-${key + 1}`}
                  value={key}
                  checked={(handler().value || []).find(
                    (item) => item.id === key
                  )}
                  onClick={(event) =>
                    updateIncidentCheckboxMulti(
                      event.target.checked,
                      value,
                      key,
                      handler().value,
                      meta,
                      parent
                    )
                  }
                />
              </Label>
            ))}
          </CheckboxGroup>
        ) : (
          <Label htmlFor={meta.name} label={meta.value} noActiveState>
            <Checkbox
              id={meta.name}
              name={meta.name}
              checked={handler().value?.value}
              onClick={(e) => {
                parent.meta.updateIncident({
                  [meta.name]: {
                    label: meta.value,
                    value: e.target.checked,
                  },
                })
              }}
            />
          </Label>
        )}
      </CheckboxWrapper>
    </FormField>
  )

CheckboxInput.defaultProps = {
  hasError: () => {},
}

CheckboxInput.propTypes = {
  handler: PropTypes.func,
  getError: PropTypes.func,
  hasError: PropTypes.func,
  meta: PropTypes.object.isRequired,
  parent: PropTypes.object,
  validatorsOrOpts: PropTypes.object,
}

export default CheckboxInput
