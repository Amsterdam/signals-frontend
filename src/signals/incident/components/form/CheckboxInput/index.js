import React from 'react';
import PropTypes from 'prop-types';

import Header from '../Header/';

const CheckboxInput = ({ handler, touched, hasError, meta, parent, getError, validatorsOrOpts }) => (
  <div className={`${meta && meta.isVisible ? 'row' : ''}`}>
    {meta && meta.isVisible ?
      <div className={`${meta.className || 'col-12'} mode_input`}>
        <Header
          meta={meta}
          options={validatorsOrOpts}
          touched={touched}
          hasError={hasError}
          getError={getError}
        >
          <div className="antwoorden">
            <div className="antwoord">
              <input
                id={meta.name}
                {...handler('checkbox')}
                onClick={(e) => parent.meta.updateIncident({ [meta.name]: e.target.checked })}
              />
              <label htmlFor={meta.name}>{meta.value}</label>
            </div>
          </div>
        </Header>
      </div>
       : ''}
  </div>
);

CheckboxInput.propTypes = {
  handler: PropTypes.func,
  touched: PropTypes.bool,
  getError: PropTypes.func,
  hasError: PropTypes.func,
  meta: PropTypes.object,
  parent: PropTypes.object,
  validatorsOrOpts: PropTypes.object
};

export default CheckboxInput;
