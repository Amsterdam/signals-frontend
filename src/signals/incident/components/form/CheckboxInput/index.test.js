import React from 'react';
import { shallow } from 'enzyme';

import CheckboxInput from './index';

describe('Form component <CheckboxInput />', () => {
  let wrapper;
  let handler;
  let touched;
  let getError;
  let hasError;
  let parent;

  beforeEach(() => {
    handler = jest.fn();
    touched = false;
    getError = jest.fn();
    hasError = jest.fn();
    parent = {
      meta: {
        updateIncident: jest.fn()
      }
    };

    wrapper = shallow(<CheckboxInput
      handler={handler}
      parent={parent}
      touched={touched}
      hasError={hasError}
      getError={getError}
    />);
  });

  describe('rendering', () => {
    it('should render checkbox correctly', () => {
      wrapper.setProps({
        meta: {
          name: 'input-field-name',
          isVisible: true
        }
      });

      expect(handler).toHaveBeenCalledWith('checkbox');
      expect(wrapper).toMatchSnapshot();
    });

    it('should render no checkbox when not visible', () => {
      wrapper.setProps({
        meta: {
          name: 'input-field-name',
          isVisible: false
        }
      });

      expect(handler).not.toHaveBeenCalled();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('events', () => {
    it('can be checked and unchecked', () => {
      wrapper.setProps({
        meta: {
          name: 'input-field-name',
          isVisible: true
        }
      });

      const checkEevent = { target: { checked: true } };
      wrapper.find('input').simulate('click', checkEevent);

      expect(parent.meta.updateIncident).toHaveBeenCalledWith({
        'input-field-name': true
      });

      const uncheckEevent = { target: { checked: false } };
      wrapper.find('input').simulate('click', uncheckEevent);

      expect(parent.meta.updateIncident).toHaveBeenCalledWith({
        'input-field-name': false
      });
    });
  });
});
