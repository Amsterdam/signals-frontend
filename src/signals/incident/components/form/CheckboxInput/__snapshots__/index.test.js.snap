// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component <CheckboxInput /> rendering should render checkbox correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-12 mode_input"
  >
    <Header
      getError={[MockFunction]}
      hasError={[MockFunction]}
      meta={
        Object {
          "isVisible": true,
          "name": "input-field-name",
        }
      }
      touched={false}
    >
      <div
        className="antwoorden"
      >
        <div
          className="antwoord"
        >
          <input
            id="input-field-name"
            onClick={[Function]}
          />
          <label
            htmlFor="input-field-name"
          />
        </div>
      </div>
    </Header>
  </div>
</div>
`;

exports[`Form component <CheckboxInput /> rendering should render no checkbox when not visible 1`] = `
<div
  className=""
/>
`;
